src/GenSymIO.chpl

```diff
@@ -258,48 +258,105 @@ module GenSymIO {
+        /**
+         * This is an H5Literate call-back function, c_helper funcs are used to process data in void*
+         * this proc counts the number of of HDF5 groups/datasets under the root, non-recursive
+         */
+        proc _get_item_count(loc_id:C_HDF5.hid_t, name:c_void_ptr, info:c_void_ptr, data:c_void_ptr) {
+            var obj_name = name:c_string;
+            var obj_type:C_HDF5.H5O_type_t;
+            var status:C_HDF5.H5O_type_t = c_get_HDF5_obj_type(loc_id, obj_name, c_ptrTo(obj_type));
+            if (obj_type == C_HDF5.H5O_TYPE_GROUP || obj_type == C_HDF5.H5O_TYPE_DATASET) {
+                c_incrementCounter(data);
             }
-            var r = f.reader(start=0);
-            r.readstring(repMsg);
-            r.close();
-        } catch e : Error {
-            var errorMsg = "failed to spawn process and execute ls: %t".format(e.message());
-            gsLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
-            return new MsgTuple(errorMsg, MsgType.ERROR);
+            return 0; // to continue iteration
         }
 
-        if exitCode != 0 {
-            var errorMsg = "could not execute ls on %s, check file permissions or format".format(filename);
-            gsLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
-            return new MsgTuple(errorMsg, MsgType.ERROR);
-        } else {
-            return new MsgTuple(repMsg, MsgType.NORMAL);
+        /**
+         * This is an H5Literate call-back function, c_helper funcs are used to process data in void*
+         * this proc builds string of HDF5 group/dataset objects names under the root, non-recursive
+         */
+        proc _simulate_h5ls(loc_id:C_HDF5.hid_t, name:c_void_ptr, info:c_void_ptr, data:c_void_ptr) {
+            var obj_name = name:c_string;
+            var obj_type:C_HDF5.H5O_type_t;
+            var status:C_HDF5.H5O_type_t = c_get_HDF5_obj_type(loc_id, obj_name, c_ptrTo(obj_type));
+            if (obj_type == C_HDF5.H5O_TYPE_GROUP || obj_type == C_HDF5.H5O_TYPE_DATASET) {
+                // items.append(obj_name:string); This doesn't work unless items is global
+                c_append_HDF5_fieldname(data, obj_name);
+            }
+            return 0; // to continue iteration
         }
+        
+        var idx_p:C_HDF5.hsize_t; // This is the H5Literate index counter
+        
+        // First iteration to get the item count so we can ballpark the char* allocation
+        var nfields:c_int = 0:c_int;
+        C_HDF5.H5Literate(fid, C_HDF5.H5_INDEX_NAME, C_HDF5.H5_INDEX_NAME, idx_p, c_ptrTo(_get_item_count), c_ptrTo(nfields));
+        
+        // Allocate space for array of strings
+        var c_field_names = c_calloc(c_char, 255 * nfields);
+        idx_p = 0:C_HDF5.hsize_t; // reset our iteration counter
+        C_HDF5.H5Literate(fid, C_HDF5.H5_INDEX_NAME, C_HDF5.H5_INDEX_NAME, idx_p, c_ptrTo(_simulate_h5ls), c_field_names);
```

Comment:
@reuster986 : I know that @glitch isn't actively developing anymore, but I noticed today that the third argument to each of the H5Literate() calls above isn't of the correct type (which the C compiler will complain about in certain modes).  Specifically, it's expecting an argument of type `H5_iter_order_t`, but we're sending it an argument of `H5_index_t`.  Chapel doesn't catch this because of the white lies that the HDF5 module uses to describe external enums, and C's permissiveness typically lets it slide as well if not compiling with various warnings turned on.  I think the argument should be switched to one of the following:

```
    // H5_iter_order_t enum                                                     
    extern type H5_iter_order_t = c_int;
    extern const H5_ITER_UNKNOWN :H5_iter_order_t;
    extern const H5_ITER_INC :H5_iter_order_t;
    extern const H5_ITER_DEC :H5_iter_order_t;
    extern const H5_ITER_NATIVE :H5_iter_order_t;
    extern const H5_ITER_N :H5_iter_order_t;
```

Reply:
I ended up opening #1189 about this in case it got lost once I closed this tab.